gh-95023: Added os.setns and os.unshare to easily switch between namespaces on Linux (#95046)
python/cpython: Doc/library/os.rst

      See the documentation for :func:`getgroups` for cases where it may not
      return the same group list set by calling setgroups().

.. function:: setns(fd, nstype=0)

   Reassociate thread with a namespace.

>    If *fd* refers to a ``/proc/[pid]/ns/`` link, :func:`setns` reassociates the

Reviewer: There is no need to link back to the function we are describing :) You can replace the `:func:setns` links with ` ``setns()`` ` in this paragraph.

   calling thread with the namespace associated with that link, subject to any

>    constraints imposed by the *nstype* argument (or any if 0).

Reviewer: ```suggestion
   constraints imposed by the *nstype* argument (or any if ``0``).
```

   Since Linux 5.8, *fd* may refer to a PID file descriptor obtained from

>    :c:func:`pidfd_open`. In this case :func:`setns` reassociates the calling thread

Reviewer: I guess you mean the `pidfd_open` syscall? In that case, the correct markup is:

```suggestion
   :manpage:`pidfd_open(2)`. In this case :func:`setns` reassociates the calling thread
```

   into one or more of the same namespaces as the thread referred to by *fd* subject

>    to any constraints imposed by the *nstype* which is a bit mask specified by ORing together one or more of the ``CLONE_NEW*`` constants. the caller's memberships in unspecified namespaces are left unchanged.

Reviewer: Could you please wrap all lines within 79 characters? Also, using semantic breaks will help reduce future diffs.

Reviewer: Also, instead of ``ORing together``, I'd suggest something like this:

```
a bit mask specified by combining one or more of the ``CLONE_NEW*`` constants using ``|`` (bitwise or).
```

Ditto for the ``ORing`` below.

>    *fd* can be any object with a :meth:`fileno()` method, or a raw file descriptor.

Reviewer: Markup tip: you don't need to add the parentheses; Sphinx will add them for you in the rendered output.

```suggestion
   *fd* can be any object with a :meth:`fileno` method, or a raw file descriptor.
```

      See the documentation for :func:`getgroups` for cases where it may not
      return the same group list set by calling setgroups().

.. function:: setns(fd, nstype=0)

>    Reassociate thread with a namespace.

Reviewer 2: ```suggestion
   Reassociate thread with a namespace, see the :manpage:`setns(2)` man page for more details.
   
```

   If *fd* refers to a ``/proc/[pid]/ns/`` link, ``setns()`` reassociates the
   calling thread with the namespace associated with that link, subject to any
   constraints imposed by the *nstype* argument (or any if ``0``).
   Since Linux 5.8, *fd* may refer to a PID file descriptor obtained from

>    :manpage:`pidfd_open(2)`. In this case ``setns()`` reassociates the calling thread

Reviewer 2: IMHO it would be better to link to our wrapper of ``pidfd_open`` than to the man page.

```
:func:`~os.pidfd_open`
```

   into one or more of the same namespaces as the thread referred to by *fd*
   subject to any constraints imposed by the *nstype*, which is
   a bit mask specified by combining one or more of the ``CLONE_NEW*`` constants

>    using ``|`` (bitwise or).

Reviewer 2: An example is probably easier to understand
```suggestion
   a bit mask specified by combining one or more of the ``CLONE_NEW*`` constants,
   e.g. ``setns(fd, os.CLONE_NEWUTS | os.CLONE_NEWPID)``.
```

or update your example to re-associate the thread with multiple new namespaces.

>    the caller's memberships in unspecified namespaces are left unchanged.

Reviewer 2: ```suggestion
   The caller's memberships in unspecified namespaces are left unchanged.
```

      See the documentation for :func:`getgroups` for cases where it may not
      return the same group list set by calling setgroups().

.. function:: setns(fd, nstype=0)

>    Reassociate thread with a namespace, see the :manpage:`setns(2)` man page for more details.

Reviewer 3: `calling thread` may be more accurate than `current thread` in my suggestion?

```suggestion
   Reassociate the current thread with a new/different namespace.
   See the :manpage:`setns(2)` man page for more details.
```

>    If *fd* refers to a ``/proc/[pid]/ns/`` link, ``setns()`` reassociates the

Reviewer 3: `setns()` should be a Sphinx cross-reference here

Reviewer: Not necessarily. We should not link to the method we are describing. This has already been discussed in the docs wg.

Reviewer 3: Aha, I missed that decision -- sorry!

So, so I know -- intra-page cross-references to *other* objects *should* be Sphinx crossrefs, but references to a function in that own function's documentation should just be code-formatted?

Reviewer: That's my understanding. And also, we don't repeat the same reference within a paragraph.

For example for the sqlite3 module docs, I avoid sqlite3 module references. Why reference the module you are describing?

   calling thread with the namespace associated with that link, subject to any

>    constraints imposed by the *nstype* argument (or any if ``0``).

Reviewer 3: Unclear what ```(or any if `0`)``` means here. I assume this is saying that there are no constraints if `nstype` is passed as zero?

If so, then something like ```(a *nstype* of `0` means no constraints)``` would be clearer to me.

   Since Linux 5.8, *fd* may refer to a PID file descriptor obtained from

>    :func:`~os.pidfd_open`. In this case ``setns()`` reassociates the calling thread

Reviewer 3: ```suggestion
   :func:`~os.pidfd_open`. In this case :func:`~os.setns()` reassociates the calling thread
```

Reviewer: Ditto

>    into one or more of the same namespaces as the thread referred to by *fd*

Reviewer 3: "one or more" -- does the caller choose which of the applicable namespaces the thread gets reassociated into? Or does it _always_ get reassociated into _all_ applicable namespaces, within the provided constraints?

Author: The next sentence is explains it: `subject to any constraints imposed by the *nstype*`

Reviewer 3: &lt;nod&gt;, I caught that constraint from the following text, but there was some ambiguity to it on my first read. On rereading, I think it's all good.

>    subject to any constraints imposed by the *nstype*, which is

Reviewer 3: This is a style preference, I would tend to write this without the "the". Perfectly fine with it, if you prefer, or if it's more typical to include, etc.

```suggestion
   subject to any constraints imposed by *nstype*, which is
```

>    a bit mask specified by combining one or more of the ``CLONE_NEW*`` constants,

Reviewer 3: You listed all of the constants in the doc for `unshare()`... would it be worth listing all of the relevant `CLONE_NEW*` constants here as well?

You also explicitly explained the bitwise-OR operator in the `unshare()` docs -- should probably either include that explanation here, or remove it from both.  (This is an audience-based decision -- is the reader likely to understand bitmasks? Given that we're talking about low-level Linux syscalls, I suspect the answer is "yes, the reader is likely to understand them" -- so, I'd be inclined to drop the explicit bitwise-OR explanation from both.

>    e.g. ``setns(fd, os.CLONE_NEWUTS | os.CLONE_NEWPID)``.

Reviewer 3: ```suggestion
   e.g., as ``setns(fd, os.CLONE_NEWUTS | os.CLONE_NEWPID)``.
```

>    The callers memberships in unspecified namespaces are left unchanged.

Reviewer 3: Or maybe `The calling thread's memberships ...`, to be really explicit about it.

```suggestion
   The caller's memberships in unspecified namespaces are left unchanged.
```

   *fd* can be any object with a :meth:`fileno` method, or a raw file descriptor.

>    This example reassociates the thread with the ``init`` process' network namespace::

Reviewer 3: ```suggestion
   This example reassociates the thread with the ``init`` process's network namespace::
```

      fd = os.open("/proc/1/ns/net", os.O_RDONLY)
      os.setns(fd, os.CLONE_NEWNET)
      os.close(fd)

>    .. availability:: Linux 3.0 or newer.

Reviewer 4: Usually there's a glibc minimum listed as well (if the LInux manpages list one, which this one does - 2.14).  Not sure if that's a doc policy, though.

   .. versionadded:: 3.12

> 

Reviewer 3: Maybe add a `.. seealso::` here, pointing down to `unshare()`?

      See the documentation for :func:`getgroups` for cases where it may not
      return the same group list set by calling setgroups().

.. function:: setns(fd, nstype=0)

>    Reassociate the current with a namespace.

Reviewer 5: ```suggestion
   Reassociate the current thread with a `Linux namespace
   <https://en.wikipedia.org/wiki/Linux_namespaces>`_.
```

* Add back missing word "thread" (I assume that's correct, from reading the `setns` manpage)?
* Add "Linux" to namespace to clarify the term and avoid confusion with its much more common use in Python
* Explicit link the term namespace to an external definition/explanation to further reduce ambiguity and provide users an easily accessible resource for a high-level summary of the main concept. Should we link something else other than what I suggested here, though?

   See the :manpage:`setns(2)` man page for more details.

>    If *fd* refers to a ``/proc/[pid]/ns/`` link, ``setns()`` reassociates the

Reviewer 5: ```suggestion
   If *fd* refers to a :file:`/proc/{pid}/ns/` link, ``setns()`` reassociates the
```

Use the appropriate role, which also gives us built-in explicit highlighting of variable portions.

   calling thread with the namespace associated with that link, subject to any
   constraints imposed by the *nstype* argument (a *nstype* of ``0`` means no
   constraints).
   Since Linux 5.8, *fd* may refer to a PID file descriptor obtained from

>    :func:`~os.pidfd_open`. In this case ``setns()`` reassociates the calling thread

Reviewer 5: ```suggestion
   :func:`~os.pidfd_open`. In this case, ``setns()`` reassociates the calling thread
```

Missing comma

>    calling thread with the namespace associated with that link, subject to any
>    constraints imposed by the *nstype* argument (a *nstype* of ``0`` means no
>    constraints).
>    Since Linux 5.8, *fd* may refer to a PID file descriptor obtained from
>    :func:`~os.pidfd_open`. In this case ``setns()`` reassociates the calling thread
>    into one or more of the same namespaces as the thread referred to by *fd*
>    subject to any constraints imposed by *nstype*, which is a bit mask specified
>    by combining one or more of the ``CLONE_NEW*`` constants,
>    e.g., as ``setns(fd, os.CLONE_NEWUTS | os.CLONE_NEWPID)``.

Reviewer 5: I personally found this passage rather confusing and hard to follow, especially given the length and density of the sentence and, in particular, the information about ``nstype`` being split up between two different cases (fd is a link vs PID file descriptor)—I was very unsure about whether the information applied to both, or only the case in question.

Assuming that it applies to both, I've reorganized this a bit to keep the sentences shorter and more focused, and avoid duplicating and splitting up the ``nstype`` information, as well as some smaller fixes and clarifications.

```suggestion
   calling thread with the namespace associated with that link. 
   Since Linux 5.8, *fd* may refer to a PID file descriptor obtained from
   :func:`~os.pidfd_open`.
   In that case, ``setns()`` reassociates the calling thread
   into one or more of the same namespaces as the thread referred to by *fd*.
   In both cases, this is subject to any constraints imposed by the *nstype*
   argument, which is a bit mask specified by
   combining one or more of the ``CLONE_NEW*`` constants,
   e.g. ``setns(fd, os.CLONE_NEWUTS | os.CLONE_NEWPID)``.
   A *nstype* of ``0``, the default, means no constraints.
```

If this is not the case, I can propose an alternate revision given clear information on which aspects of `nstype` apply to which.

Author: `nstype` is not a bit mask in both cases. In the the first case (a link to a `/proc/pid/ns`), nstype can have one of the constants as value, in the second case (PID fd) it can be a bitmask of many namespaces to enter

Reviewer 5: I see, thanks; it is even more unclear to me that such was the intent, then, and the valid values in the first case are never explicitly stated at all. As mentioned above, I will propose a revision to clarify it accordingly along with incorporating the other fixes above.

   The caller's memberships in unspecified namespaces are left unchanged.

>    *fd* can be any object with a :meth:`fileno` method, or a raw file descriptor.

Reviewer 5: I suggest either linking to an example of the appropriate method:

```suggestion
   *fd* can be any object with a :meth:`~io.~IOBase.fileno` method, or a raw file descriptor.
```

or not making this a (broken) reference at all:

```suggestion
   *fd* can be any object with a ``fileno`` method, or a raw file descriptor.
```

      See the documentation for :func:`getgroups` for cases where it may not
      return the same group list set by calling setgroups().

.. function:: setns(fd, nstype=0)

   Reassociate the current thread with a Linux namespace.
   See the :manpage:`setns(2)` and :manpage:`namespaces(7)` man pages for more
   details.

   If *fd* refers to a :file:`/proc/{pid}/ns/` link, ``setns()`` reassociates the
   calling thread with the namespace associated with that link, subject to any
   constraints imposed by the *nstype* argument (a *nstype* of ``0`` means no

>    constraints).

Reviewer 5: ```suggestion
   If *fd* refers to a :file:`/proc/{pid}/ns/` link, ``setns()`` reassociates the
   calling thread with the namespace associated with that link,
   and *nstype* may be set to one of the
   :ref:`CLONE_NEW* constants <os-unshare-clone-flags>`
   to impose constraints on the operation
   (``0`` means no constraints).
```

* Clearly state and link what values `nstype` may be set to in order to impose constraints (assuming I now understand it correctly based on your helpful comment).
* Revise structure to avoid overly long and hard to follow syntax

   Since Linux 5.8, *fd* may refer to a PID file descriptor obtained from
   :func:`~os.pidfd_open`. In this case, ``setns()`` reassociates the calling thread
   into one or more of the same namespaces as the thread referred to by *fd*
   subject to any constraints imposed by *nstype*, which is a bit mask specified
   by combining one or more of the ``CLONE_NEW*`` constants,

>    e.g., as ``setns(fd, os.CLONE_NEWUTS | os.CLONE_NEWPID)``.

Reviewer 5: ```suggestion

   Since Linux 5.8, *fd* may refer to a PID file descriptor obtained from
   :func:`~os.pidfd_open`. In this case, ``setns()`` reassociates the calling thread
   into one or more of the same namespaces as the thread referred to by *fd*.
   This is subject to any constraints imposed by *nstype*,
   which is is a bit mask combining one or more of the
   :ref:`CLONE_NEW* constants <os-unshare-clone-flags>`,
   e.g. ``setns(fd, os.CLONE_NEWUTS | os.CLONE_NEWPID)``.
```

* Add a paragraph break to make seperation of the two cases clearer and break up a relatively long paragraph
* Split up into two sentences, to avoid one overly long and hard to follow one.
* Fix grammar/punctuation issues in final sentence and slightly simplify
* Link the `CLONE_NEW*` constants

>    The caller's memberships in unspecified namespaces are left unchanged.
>    *fd* can be any object with a :meth:`~io.IOBase.fileno` method, or a raw file descriptor.

Reviewer 5: Are these sentences specific to the second case (I'm guessing maybe yes?), or do one or both apply to either? It is left unclear by the previous structure; as it stands being part of the newly-split second paragraph they apply to only the 
latter, but if one or both applies to both, then they should be separated into their own paragraph:

```suggestion

   The caller's memberships in unspecified namespaces are left unchanged.
   *fd* can be any object with a :meth:`~io.IOBase.fileno` method, or a raw file descriptor.
```

or 

```suggestion
   The caller's memberships in unspecified namespaces are left unchanged.

   *fd* can be any object with a :meth:`~io.IOBase.fileno` method, or a raw file descriptor.
```

If they both apply only to the second case, then this should be left as-is.

.. function:: setns(fd, nstype=0)

   Reassociate thread with a namespace.
   If *fd* refers to a ``/proc/[pid]/ns/`` link, ``setns()`` reassociates the
   calling thread with the namespace associated with that link, subject to any
   constraints imposed by the *nstype* argument (or any if ``0``).
   Since Linux 5.8, *fd* may refer to a PID file descriptor obtained from
   :manpage:`pidfd_open(2)`. In this case ``setns()`` reassociates the calling thread
   into one or more of the same namespaces as the thread referred to by *fd*
   subject to any constraints imposed by the *nstype*, which is
   a bit mask specified by combining one or more of the ``CLONE_NEW*`` constants
   using ``|`` (bitwise or).

>    the caller's memberships in unspecified namespaces are left unchanged.

Reviewer: ```suggestion
   The callers memberships in unspecified namespaces are left unchanged.
```

      The function is now always available and is also available on Windows.

.. function:: unshare(flags)

   Disassociate parts of the process execution context.
   The *flags* argument is a bit mask that specifies which parts of the execution
   context should be unshared. This argument is specified by ORing together zero

>    or more of the ``CLONE_*`` constants.

Reviewer: I think these sentences could be combined.

>    If *flags* is specified as zero, no changes are made to the calling process's

Reviewer: process's => process'

      The function is now always available and is also available on Windows.

.. function:: unshare(flags)

>    Disassociate parts of the process execution context.

Reviewer 2: Also link to the man page here.

>    The *flags* argument is a bit mask combining zero or more of the ``CLONE_*`` constants using ``|`` (bitwise or), that specifies which parts of the execution

Reviewer 2: Other places in our documentation use the phrase ``bitwise ORed combinations of``

Reviewer: And other places uses wordings like this. FTR, I got that recommendation from the proofreaders team who are far better in technical writing than I am :)

   context should be unshared.
   If *flags* is specified as zero, no changes are made to the calling process'
   execution context.

   .. availability:: Linux 2.6.16 or newer.

   .. versionadded:: 3.12

   .. seealso::

>       The :func:`.setns` function.

Reviewer 2: ```suggestion
      The :func:`~os.setns` function.
```

      The function is now always available and is also available on Windows.

.. function:: unshare(flags)

>    Disassociate parts of the process execution context, see the :manpage:`unshare(2)`

Reviewer 3: This line could probably use to talk about namespaces more explicitly, I think -- at the very least, it would be good for it to use the word 'namespace' (or 'namespaces') somewhere here. 

```suggestion
   Disassociate parts of the process execution context and move them into a newly created namespace.
   See the :manpage:`unshare(2)`
```

   man page for more details.
   The *flags* argument is a bit mask combining zero or more of the ``CLONE_*``
   constants using ``|`` (bitwise or), that specifies which parts of the execution

>    context should be unshared.

Reviewer 3: ```suggestion
   context should be unshared from their existing associations and moved to a new namespace.
```

>    If *flags* is specified as zero, no changes are made to the calling process'

Reviewer 3: ```suggestion
   If *flags* is specified as zero, no changes are made to the calling process's
```

Reviewer: Ah, TIL :)

Reviewer 3: @erlend-aasland Yeah, there's some controversy about this, and my understanding might be out of date, but AFAIK if a thing is singular, you still always put the `'s` to indicate possession, even if it ends with `s`. You only have the `s'` when the thing is plural.

It leads to some wacky things ... like, the house that belongs to the Jones family is "the Joneses' house", whereas the house that belongs to Bill Jones is "Jones's house"... :man_shrugging: :boom:

.. function:: unshare(flags)

   Disassociate parts of the process execution context.

>    The *flags* argument is a bit mask combining zero or more of the ``CLONE_*`` constants using ``|`` (bitwise or), that specifies which parts of the execution

Reviewer: Please wrap to 79 char line width.

      The function is now always available and is also available on Windows.

.. function:: unshare(flags)

   Disassociate parts of the process execution context, and move them into a
   newly created namespace.
   See the :manpage:`unshare(2)`
   man page for more details.

>    The *flags* argument is a bit mask combining zero or more of the ``CLONE_*``

Reviewer 5: ```suggestion
   The *flags* argument is a bit mask, combining zero or more of the ``CLONE_*``
```

Set off this clause with commas to make it a little easier to follow the flow

>    constants using ``|`` (bitwise or), that specifies which parts of the execution

Reviewer 5: Maybe link to the actual operator?

```suggestion
   constants using :func:` | (bitwise or) <operator.or_>`,
   that specifies which parts of the execution
```

   context should be unshared from their existing associations and moved to a

>    new namespace..

Reviewer 5: ```suggestion
   new namespace.
```

Fix typo

>    If *flags* is specified as zero, no changes are made to the calling process's

Reviewer 5: ```suggestion
   If the *flags* argument is ``0``, no changes are made to the calling process's
```

Use an actual literal where appropriate, avoid awkwardness over flags being singular and eliminate extra wordiness

   execution context.

   .. availability:: Linux 2.6.16 or newer.

   .. versionadded:: 3.12

   .. seealso::

      The :func:`~os.setns` function.

Flags to the :func:`unshare` function, if the implementation supports them.

> See the Linux manual for the exact effect and availability.

Reviewer 5: ```suggestion
See :manpage:`clone(2)` in the Linux manual
for their exact effect and availability.
```

Explicitly state and link the actual manual entry; I had to go digging to find it.

.. data:: CLONE_FILES
          CLONE_FS
          CLONE_NEWCGROUP
          CLONE_NEWIPC
          CLONE_NEWNET
          CLONE_NEWNS
          CLONE_NEWPID
          CLONE_NEWTIME
          CLONE_NEWUSER
          CLONE_NEWUTS
          CLONE_SIGHAND
          CLONE_SYSVSEM
          CLONE_THREAD

>           CLONE_VM
> 
> .. _os-newstreams:

Reviewer 5: ```suggestion
          CLONE_VM

.. _os-newstreams:
```

Missing line break

      The function is now always available and is also available on Windows.

.. function:: unshare(flags)

   Disassociate parts of the process execution context, and move them into a
   newly created namespace.
   See the :manpage:`unshare(2)`
   man page for more details.
   The *flags* argument is a bit mask, combining zero or more of the ``CLONE_*``
   constants, that specifies which parts of the execution context should be
   unshared from their existing associations and moved to a new namespace.
   If the *flags* argument is ``0``, no changes are made to the calling process's
   execution context.

   .. availability:: Linux 2.6.16 or newer.

   .. versionadded:: 3.12

   .. seealso::

      The :func:`~os.setns` function.

> 

Reviewer 5: ```suggestion

.. _os-unshare-clone-flags:

```

Add to enable inbound references to the `CLONE_*` flags as a whole, as needed above.

Flags to the :func:`unshare` function, if the implementation supports them.
See the Linux manual for the exact effect and availability.

.. data:: CLONE_FILES
          CLONE_NEWCGROUP
          CLONE_NEWIPC
          CLONE_NEWNET
          CLONE_NEWNS
          CLONE_NEWPID
          CLONE_NEWTIME
          CLONE_NEWUSER
          CLONE_NEWUTS
          CLONE_SIGHAND
          CLONE_SYSVSEM
          CLONE_THREAD
          CLONE_VM

>           CLONE_FS

Reviewer: Please sort this list alphabetically :)
